// mendixmodelsdk/dist/gen/exportmappings.js
"use strict";
/* tslint:disable */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var mappings_1 = require("./mappings");
var exportmappings;
(function (exportmappings) {
    /**
     * Interfaces and instance classes for types from the Mendix sub meta model `ExportMappings`.
     */
    /**
     * See: {@link https://docs.mendix.com/refguide7/export-mappings relevant section in reference guide}
     */
    var ExportMapping = /** @class */ (function (_super) {
        __extends(ExportMapping, _super);
        function ExportMapping(model, structureTypeName, id, isPartial, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, container) || this;
            /** @internal */
            _this.__parameterName = new internal.PrimitiveProperty(ExportMapping, _this, "parameterName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__parameterTypeName = new internal.PrimitiveProperty(ExportMapping, _this, "parameterTypeName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__isHeader = new internal.PrimitiveProperty(ExportMapping, _this, "isHeader", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__nullValueOption = new internal.EnumProperty(ExportMapping, _this, "nullValueOption", microflows_1.microflows.NullValueOption.LeaveOutElement, microflows_1.microflows.NullValueOption);
            _this._containmentName = "documents";
            return _this;
        }
        Object.defineProperty(ExportMapping.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ExportMapping.prototype, "parameterName", {
            get: function () { return this.__parameterName.get(); },
            set: function (newValue) { this.__parameterName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ExportMapping.prototype, "parameterTypeName", {
            /**
             * In version 6.1.0: deleted
             */
            get: function () { return this.__parameterTypeName.get(); },
            set: function (newValue) { this.__parameterTypeName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ExportMapping.prototype, "isHeader", {
            get: function () { return this.__isHeader.get(); },
            set: function (newValue) { this.__isHeader.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ExportMapping.prototype, "nullValueOption", {
            /**
             * In version 6.7.0: introduced
             */
            get: function () { return this.__nullValueOption.get(); },
            set: function (newValue) { this.__nullValueOption.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates a new ExportMapping unit in the SDK and on the server.
         * Expects one argument, the projects.IFolderBase in which this unit is contained.
         */
        ExportMapping.createIn = function (container) {
            return internal.instancehelpers.createUnit(container, ExportMapping);
        };
        /** @internal */
        ExportMapping.prototype._isByNameReferrable = function () { return true; };
        /** @internal */
        ExportMapping.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            if (this.__nullValueOption.isAvailable) {
                this.nullValueOption = microflows_1.microflows.NullValueOption.LeaveOutElement;
            }
        };
        ExportMapping.structureTypeName = "ExportMappings$ExportMapping";
        ExportMapping.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "parameterTypeName": {
                    "deleted": "6.1.0",
                    "deletionMessage": null
                },
                "nullValueOption": {
                    "introduced": "6.7.0"
                }
            }
        }, internal.StructureType.ModelUnit);
        return ExportMapping;
    }(mappings_1.mappings.MappingDocument));
    exportmappings.ExportMapping = ExportMapping;
    /**
     * See: {@link https://docs.mendix.com/refguide7/export-mappings relevant section in reference guide}
     */
    var ExportObjectMappingElement = /** @class */ (function (_super) {
        __extends(ExportObjectMappingElement, _super);
        function ExportObjectMappingElement(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new ExportObjectMappingElement() cannot be invoked directly, please use 'model.exportmappings.createExportObjectMappingElement()'");
            }
            return _this;
        }
        Object.defineProperty(ExportObjectMappingElement.prototype, "containerAsMappingDocument", {
            get: function () { return _super.prototype.getContainerAs.call(this, mappings_1.mappings.MappingDocument); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ExportObjectMappingElement.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, mappings_1.mappings.ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new ExportObjectMappingElement instance in the SDK and on the server.
         * The new ExportObjectMappingElement will be automatically stored in the 'rootMappingElements' property
         * of the parent mappings.MappingDocument element passed as argument.
         */
        ExportObjectMappingElement.createInMappingDocumentUnderRootMappingElements = function (container) {
            return internal.instancehelpers.createElement(container, ExportObjectMappingElement, "rootMappingElements", true);
        };
        /**
         * Creates and returns a new ExportObjectMappingElement instance in the SDK and on the server.
         * The new ExportObjectMappingElement will be automatically stored in the 'children' property
         * of the parent mappings.ObjectMappingElement element passed as argument.
         */
        ExportObjectMappingElement.createInObjectMappingElementUnderChildren = function (container) {
            return internal.instancehelpers.createElement(container, ExportObjectMappingElement, "children", true);
        };
        /**
         * Creates and returns a new ExportObjectMappingElement instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        ExportObjectMappingElement.create = function (model) {
            return internal.instancehelpers.createElement(model, ExportObjectMappingElement);
        };
        /** @internal */
        ExportObjectMappingElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        ExportObjectMappingElement.structureTypeName = "ExportMappings$ExportObjectMappingElement";
        ExportObjectMappingElement.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
        return ExportObjectMappingElement;
    }(mappings_1.mappings.ObjectMappingElement));
    exportmappings.ExportObjectMappingElement = ExportObjectMappingElement;
    /**
     * See: {@link https://docs.mendix.com/refguide7/export-mappings relevant section in reference guide}
     */
    var ExportValueMappingElement = /** @class */ (function (_super) {
        __extends(ExportValueMappingElement, _super);
        function ExportValueMappingElement(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new ExportValueMappingElement() cannot be invoked directly, please use 'model.exportmappings.createExportValueMappingElement()'");
            }
            return _this;
        }
        Object.defineProperty(ExportValueMappingElement.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, mappings_1.mappings.ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new ExportValueMappingElement instance in the SDK and on the server.
         * The new ExportValueMappingElement will be automatically stored in the 'children' property
         * of the parent mappings.ObjectMappingElement element passed as argument.
         */
        ExportValueMappingElement.createIn = function (container) {
            return internal.instancehelpers.createElement(container, ExportValueMappingElement, "children", true);
        };
        /**
         * Creates and returns a new ExportValueMappingElement instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        ExportValueMappingElement.create = function (model) {
            return internal.instancehelpers.createElement(model, ExportValueMappingElement);
        };
        /** @internal */
        ExportValueMappingElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        ExportValueMappingElement.structureTypeName = "ExportMappings$ExportValueMappingElement";
        ExportValueMappingElement.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
        return ExportValueMappingElement;
    }(mappings_1.mappings.ValueMappingElement));
    exportmappings.ExportValueMappingElement = ExportValueMappingElement;
})(exportmappings = exports.exportmappings || (exports.exportmappings = {}));
var microflows_1 = require("./microflows");
var projects_1 = require("./projects");
