// mendixmodelsdk/dist/gen/mappings.js
"use strict";
/* tslint:disable */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var projects_1 = require("./projects");
var mappings;
(function (mappings) {
    var ElementType = /** @class */ (function (_super) {
        __extends(ElementType, _super);
        function ElementType() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "mappings.ElementType";
            return _this;
        }
        ElementType.Undefined = new ElementType("Undefined", {});
        ElementType.Inheritance = new ElementType("Inheritance", {});
        ElementType.Choice = new ElementType("Choice", {});
        ElementType.Object = new ElementType("Object", {});
        ElementType.Value = new ElementType("Value", {});
        ElementType.Sequence = new ElementType("Sequence", {});
        ElementType.All = new ElementType("All", {});
        ElementType.NamedArray = new ElementType("NamedArray", {
            "introduced": "6.4.0",
            "deleted": "6.6.0",
            "deletionMessage": null
        });
        ElementType.Array = new ElementType("Array", {
            "introduced": "6.6.0"
        });
        ElementType.Wrapper = new ElementType("Wrapper", {
            "introduced": "6.4.0"
        });
        return ElementType;
    }(internal.AbstractEnum));
    mappings.ElementType = ElementType;
    var ObjectHandlingBackupEnum = /** @class */ (function (_super) {
        __extends(ObjectHandlingBackupEnum, _super);
        function ObjectHandlingBackupEnum() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "mappings.ObjectHandlingBackupEnum";
            return _this;
        }
        ObjectHandlingBackupEnum.Create = new ObjectHandlingBackupEnum("Create", {});
        ObjectHandlingBackupEnum.Ignore = new ObjectHandlingBackupEnum("Ignore", {});
        ObjectHandlingBackupEnum.Error = new ObjectHandlingBackupEnum("Error", {});
        return ObjectHandlingBackupEnum;
    }(internal.AbstractEnum));
    mappings.ObjectHandlingBackupEnum = ObjectHandlingBackupEnum;
    var ObjectHandlingEnum = /** @class */ (function (_super) {
        __extends(ObjectHandlingEnum, _super);
        function ObjectHandlingEnum() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "mappings.ObjectHandlingEnum";
            return _this;
        }
        ObjectHandlingEnum.Parameter = new ObjectHandlingEnum("Parameter", {
            "introduced": "7.5.0"
        });
        ObjectHandlingEnum.Create = new ObjectHandlingEnum("Create", {});
        ObjectHandlingEnum.Find = new ObjectHandlingEnum("Find", {});
        ObjectHandlingEnum.Custom = new ObjectHandlingEnum("Custom", {});
        return ObjectHandlingEnum;
    }(internal.AbstractEnum));
    mappings.ObjectHandlingEnum = ObjectHandlingEnum;
    /**
     * Interfaces and instance classes for types from the Mendix sub meta model `Mappings`.
     */
    /**
     * In version 6.6.0: introduced
     */
    var Element = /** @class */ (function (_super) {
        __extends(Element, _super);
        function Element(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__elementType = new internal.EnumProperty(Element, _this, "elementType", ElementType.Value, ElementType);
            /** @internal */
            _this.__primitiveType = new internal.EnumProperty(Element, _this, "primitiveType", xmlschemas_1.xmlschemas.XmlPrimitiveType.Unknown, xmlschemas_1.xmlschemas.XmlPrimitiveType);
            /** @internal */
            _this.__path = new internal.PrimitiveProperty(Element, _this, "path", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__isDefaultType = new internal.PrimitiveProperty(Element, _this, "isDefaultType", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__minOccurs = new internal.PrimitiveProperty(Element, _this, "minOccurs", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__maxOccurs = new internal.PrimitiveProperty(Element, _this, "maxOccurs", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__nillable = new internal.PrimitiveProperty(Element, _this, "nillable", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__exposedName = new internal.PrimitiveProperty(Element, _this, "exposedName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__exposedItemName = new internal.PrimitiveProperty(Element, _this, "exposedItemName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__maxLength = new internal.PrimitiveProperty(Element, _this, "maxLength", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__fractionDigits = new internal.PrimitiveProperty(Element, _this, "fractionDigits", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__totalDigits = new internal.PrimitiveProperty(Element, _this, "totalDigits", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__errorMessage = new internal.PrimitiveProperty(Element, _this, "errorMessage", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__warningMessage = new internal.PrimitiveProperty(Element, _this, "warningMessage", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__children = new internal.PartListProperty(Element, _this, "children", []);
            if (arguments.length < 4) {
                throw new Error("new Element() cannot be invoked directly, please use 'model.mappings.createElement()'");
            }
            return _this;
        }
        Object.defineProperty(Element.prototype, "containerAsJsonStructure", {
            get: function () { return _super.prototype.getContainerAs.call(this, jsonstructures_1.jsonstructures.JsonStructure); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "containerAsElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, Element); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "containerAsEntityMessageDefinition", {
            get: function () { return _super.prototype.getContainerAs.call(this, messagedefinitions_1.messagedefinitions.EntityMessageDefinition); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "elementType", {
            get: function () { return this.__elementType.get(); },
            set: function (newValue) { this.__elementType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "primitiveType", {
            get: function () { return this.__primitiveType.get(); },
            set: function (newValue) { this.__primitiveType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "path", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             */
            get: function () { return this.__path.get(); },
            set: function (newValue) { this.__path.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "isDefaultType", {
            get: function () { return this.__isDefaultType.get(); },
            set: function (newValue) { this.__isDefaultType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "minOccurs", {
            get: function () { return this.__minOccurs.get(); },
            set: function (newValue) { this.__minOccurs.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "maxOccurs", {
            get: function () { return this.__maxOccurs.get(); },
            set: function (newValue) { this.__maxOccurs.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "nillable", {
            get: function () { return this.__nillable.get(); },
            set: function (newValue) { this.__nillable.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "exposedName", {
            get: function () { return this.__exposedName.get(); },
            set: function (newValue) { this.__exposedName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "exposedItemName", {
            /**
             * In version 7.6.0: introduced
             */
            get: function () { return this.__exposedItemName.get(); },
            set: function (newValue) { this.__exposedItemName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "maxLength", {
            get: function () { return this.__maxLength.get(); },
            set: function (newValue) { this.__maxLength.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "fractionDigits", {
            get: function () { return this.__fractionDigits.get(); },
            set: function (newValue) { this.__fractionDigits.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "totalDigits", {
            get: function () { return this.__totalDigits.get(); },
            set: function (newValue) { this.__totalDigits.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "errorMessage", {
            get: function () { return this.__errorMessage.get(); },
            set: function (newValue) { this.__errorMessage.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "warningMessage", {
            get: function () { return this.__warningMessage.get(); },
            set: function (newValue) { this.__warningMessage.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Element.prototype, "children", {
            get: function () { return this.__children.get(); },
            enumerable: true,
            configurable: true
        });
        /** @internal */
        Element.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.elementType = ElementType.Value;
            this.fractionDigits = -1;
            this.maxLength = -1;
            this.maxOccurs = 1;
            this.minOccurs = 1;
            this.nillable = false;
            this.primitiveType = xmlschemas_1.xmlschemas.XmlPrimitiveType.Unknown;
            this.totalDigits = -1;
        };
        Element.structureTypeName = "Mappings$Element";
        Element.versionInfo = new exports.StructureVersionInfo({
            "introduced": "6.6.0",
            "properties": {
                "exposedItemName": {
                    "introduced": "7.6.0"
                }
            }
        }, internal.StructureType.Element);
        return Element;
    }(internal.Element));
    mappings.Element = Element;
    /**
     * See: {@link https://docs.mendix.com/refguide7/mapping-documents relevant section in reference guide}
     */
    var MappingDocument = /** @class */ (function (_super) {
        __extends(MappingDocument, _super);
        function MappingDocument(model, structureTypeName, id, isPartial, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, container) || this;
            /** @internal */
            _this.__rootMappingElements = new internal.PartListProperty(MappingDocument, _this, "rootMappingElements", []);
            /** @internal */
            _this.__xmlSchema = new internal.ByNameReferenceProperty(MappingDocument, _this, "xmlSchema", null, "XmlSchemas$XmlSchema");
            /** @internal */
            _this.__jsonStructure = new internal.ByNameReferenceProperty(MappingDocument, _this, "jsonStructure", null, "JsonStructures$JsonStructure");
            /** @internal */
            _this.__rootElementName = new internal.PrimitiveProperty(MappingDocument, _this, "rootElementName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__importedWebService = new internal.ByNameReferenceProperty(MappingDocument, _this, "importedWebService", null, "WebServices$ImportedWebService");
            /** @internal */
            _this.__serviceName = new internal.PrimitiveProperty(MappingDocument, _this, "serviceName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__operationName = new internal.PrimitiveProperty(MappingDocument, _this, "operationName", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__messageDefinition = new internal.ByNameReferenceProperty(MappingDocument, _this, "messageDefinition", null, "MessageDefinitions$MessageDefinition");
            return _this;
        }
        Object.defineProperty(MappingDocument.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "rootMappingElements", {
            get: function () { return this.__rootMappingElements.get(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "xmlSchema", {
            get: function () { return this.__xmlSchema.get(); },
            set: function (newValue) { this.__xmlSchema.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "xmlSchemaQualifiedName", {
            get: function () { return this.__xmlSchema.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "jsonStructure", {
            /**
             * In version 6.4.0: introduced
             */
            get: function () { return this.__jsonStructure.get(); },
            set: function (newValue) { this.__jsonStructure.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "jsonStructureQualifiedName", {
            get: function () { return this.__jsonStructure.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "rootElementName", {
            get: function () { return this.__rootElementName.get(); },
            set: function (newValue) { this.__rootElementName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "importedWebService", {
            get: function () { return this.__importedWebService.get(); },
            set: function (newValue) { this.__importedWebService.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "importedWebServiceQualifiedName", {
            get: function () { return this.__importedWebService.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "serviceName", {
            get: function () { return this.__serviceName.get(); },
            set: function (newValue) { this.__serviceName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "operationName", {
            get: function () { return this.__operationName.get(); },
            set: function (newValue) { this.__operationName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "messageDefinition", {
            /**
             * In version 7.6.0: introduced
             */
            get: function () { return this.__messageDefinition.get(); },
            set: function (newValue) { this.__messageDefinition.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingDocument.prototype, "messageDefinitionQualifiedName", {
            get: function () { return this.__messageDefinition.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /** @internal */
        MappingDocument.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        MappingDocument.structureTypeName = "Mappings$MappingDocument";
        MappingDocument.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "jsonStructure": {
                    "introduced": "6.4.0"
                },
                "messageDefinition": {
                    "introduced": "7.6.0"
                }
            }
        }, internal.StructureType.ModelUnit);
        return MappingDocument;
    }(projects_1.projects.Document));
    mappings.MappingDocument = MappingDocument;
    var MappingElement = /** @class */ (function (_super) {
        __extends(MappingElement, _super);
        function MappingElement(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__documentation = new internal.PrimitiveProperty(MappingElement, _this, "documentation", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__elementType = new internal.EnumProperty(MappingElement, _this, "elementType", ElementType.Undefined, ElementType);
            /** @internal */
            _this.__path = new internal.PrimitiveProperty(MappingElement, _this, "path", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__jsonPath = new internal.PrimitiveProperty(MappingElement, _this, "jsonPath", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__xmlPath = new internal.PrimitiveProperty(MappingElement, _this, "xmlPath", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__minOccurs = new internal.PrimitiveProperty(MappingElement, _this, "minOccurs", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__maxOccurs = new internal.PrimitiveProperty(MappingElement, _this, "maxOccurs", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__nillable = new internal.PrimitiveProperty(MappingElement, _this, "nillable", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__exposedName = new internal.PrimitiveProperty(MappingElement, _this, "exposedName", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new MappingElement() cannot be invoked directly, please use 'model.mappings.createMappingElement()'");
            }
            return _this;
        }
        Object.defineProperty(MappingElement.prototype, "containerAsMappingDocument", {
            get: function () { return _super.prototype.getContainerAs.call(this, MappingDocument); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "documentation", {
            get: function () { return this.__documentation.get(); },
            set: function (newValue) { this.__documentation.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "elementType", {
            get: function () { return this.__elementType.get(); },
            set: function (newValue) { this.__elementType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "path", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: deleted
             */
            get: function () { return this.__path.get(); },
            set: function (newValue) { this.__path.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "jsonPath", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: introduced
             */
            get: function () { return this.__jsonPath.get(); },
            set: function (newValue) { this.__jsonPath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "xmlPath", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: introduced
             */
            get: function () { return this.__xmlPath.get(); },
            set: function (newValue) { this.__xmlPath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "minOccurs", {
            get: function () { return this.__minOccurs.get(); },
            set: function (newValue) { this.__minOccurs.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "maxOccurs", {
            get: function () { return this.__maxOccurs.get(); },
            set: function (newValue) { this.__maxOccurs.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "nillable", {
            get: function () { return this.__nillable.get(); },
            set: function (newValue) { this.__nillable.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingElement.prototype, "exposedName", {
            /**
             * In version 6.6.0: introduced
             */
            get: function () { return this.__exposedName.get(); },
            set: function (newValue) { this.__exposedName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /** @internal */
        MappingElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.elementType = ElementType.Undefined;
        };
        MappingElement.structureTypeName = "Mappings$MappingElement";
        MappingElement.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "path": {
                    "deleted": "7.6.0",
                    "deletionMessage": null
                },
                "jsonPath": {
                    "introduced": "7.6.0"
                },
                "xmlPath": {
                    "introduced": "7.6.0"
                },
                "exposedName": {
                    "introduced": "6.6.0"
                }
            }
        }, internal.StructureType.Element);
        return MappingElement;
    }(internal.Element));
    mappings.MappingElement = MappingElement;
    var MappingMicroflowCall = /** @class */ (function (_super) {
        __extends(MappingMicroflowCall, _super);
        function MappingMicroflowCall(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__microflow = new internal.ByNameReferenceProperty(MappingMicroflowCall, _this, "microflow", null, "Microflows$Microflow");
            /** @internal */
            _this.__parameterMappings = new internal.PartListProperty(MappingMicroflowCall, _this, "parameterMappings", []);
            if (arguments.length < 4) {
                throw new Error("new MappingMicroflowCall() cannot be invoked directly, please use 'model.mappings.createMappingMicroflowCall()'");
            }
            return _this;
        }
        Object.defineProperty(MappingMicroflowCall.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowCall.prototype, "microflow", {
            get: function () { return this.__microflow.get(); },
            set: function (newValue) { this.__microflow.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowCall.prototype, "microflowQualifiedName", {
            get: function () { return this.__microflow.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowCall.prototype, "parameterMappings", {
            get: function () { return this.__parameterMappings.get(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new MappingMicroflowCall instance in the SDK and on the server.
         * The new MappingMicroflowCall will be automatically stored in the 'mappingMicroflowCall' property
         * of the parent ObjectMappingElement element passed as argument.
         */
        MappingMicroflowCall.createIn = function (container) {
            return internal.instancehelpers.createElement(container, MappingMicroflowCall, "mappingMicroflowCall", false);
        };
        /**
         * Creates and returns a new MappingMicroflowCall instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        MappingMicroflowCall.create = function (model) {
            return internal.instancehelpers.createElement(model, MappingMicroflowCall);
        };
        /** @internal */
        MappingMicroflowCall.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        MappingMicroflowCall.structureTypeName = "Mappings$MappingMicroflowCall";
        MappingMicroflowCall.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
        return MappingMicroflowCall;
    }(internal.Element));
    mappings.MappingMicroflowCall = MappingMicroflowCall;
    var MappingMicroflowParameter = /** @class */ (function (_super) {
        __extends(MappingMicroflowParameter, _super);
        function MappingMicroflowParameter(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__parameter = new internal.ByNameReferenceProperty(MappingMicroflowParameter, _this, "parameter", null, "Microflows$MicroflowParameter");
            /** @internal */
            _this.__levelOfParent = new internal.PrimitiveProperty(MappingMicroflowParameter, _this, "levelOfParent", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__valueElementPath = new internal.PrimitiveProperty(MappingMicroflowParameter, _this, "valueElementPath", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__jsonValueElementPath = new internal.PrimitiveProperty(MappingMicroflowParameter, _this, "jsonValueElementPath", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__xmlValueElementPath = new internal.PrimitiveProperty(MappingMicroflowParameter, _this, "xmlValueElementPath", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new MappingMicroflowParameter() cannot be invoked directly, please use 'model.mappings.createMappingMicroflowParameter()'");
            }
            return _this;
        }
        Object.defineProperty(MappingMicroflowParameter.prototype, "containerAsMappingMicroflowCall", {
            get: function () { return _super.prototype.getContainerAs.call(this, MappingMicroflowCall); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "parameter", {
            get: function () { return this.__parameter.get(); },
            set: function (newValue) { this.__parameter.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "parameterQualifiedName", {
            get: function () { return this.__parameter.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "levelOfParent", {
            get: function () { return this.__levelOfParent.get(); },
            set: function (newValue) { this.__levelOfParent.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "valueElementPath", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: deleted
             */
            get: function () { return this.__valueElementPath.get(); },
            set: function (newValue) { this.__valueElementPath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "jsonValueElementPath", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: introduced
             */
            get: function () { return this.__jsonValueElementPath.get(); },
            set: function (newValue) { this.__jsonValueElementPath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MappingMicroflowParameter.prototype, "xmlValueElementPath", {
            /**
             * The value of this property is conceptually of type mappings.ElementPath.
             *
             * In version 7.6.0: introduced
             */
            get: function () { return this.__xmlValueElementPath.get(); },
            set: function (newValue) { this.__xmlValueElementPath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new MappingMicroflowParameter instance in the SDK and on the server.
         * The new MappingMicroflowParameter will be automatically stored in the 'parameterMappings' property
         * of the parent MappingMicroflowCall element passed as argument.
         */
        MappingMicroflowParameter.createIn = function (container) {
            return internal.instancehelpers.createElement(container, MappingMicroflowParameter, "parameterMappings", true);
        };
        /**
         * Creates and returns a new MappingMicroflowParameter instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        MappingMicroflowParameter.create = function (model) {
            return internal.instancehelpers.createElement(model, MappingMicroflowParameter);
        };
        /** @internal */
        MappingMicroflowParameter.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.levelOfParent = -1;
        };
        MappingMicroflowParameter.structureTypeName = "Mappings$MappingMicroflowParameter";
        MappingMicroflowParameter.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "valueElementPath": {
                    "deleted": "7.6.0",
                    "deletionMessage": null
                },
                "jsonValueElementPath": {
                    "introduced": "7.6.0"
                },
                "xmlValueElementPath": {
                    "introduced": "7.6.0"
                }
            }
        }, internal.StructureType.Element);
        return MappingMicroflowParameter;
    }(internal.Element));
    mappings.MappingMicroflowParameter = MappingMicroflowParameter;
    var ObjectMappingElement = /** @class */ (function (_super) {
        __extends(ObjectMappingElement, _super);
        function ObjectMappingElement(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__mappingMicroflowCall = new internal.PartProperty(ObjectMappingElement, _this, "mappingMicroflowCall", null, false);
            /** @internal */
            _this.__children = new internal.PartListProperty(ObjectMappingElement, _this, "children", []);
            /** @internal */
            _this.__entity = new internal.ByNameReferenceProperty(ObjectMappingElement, _this, "entity", null, "DomainModels$Entity");
            /** @internal */
            _this.__association = new internal.ByNameReferenceProperty(ObjectMappingElement, _this, "association", null, "DomainModels$AssociationBase");
            /** @internal */
            _this.__objectHandling = new internal.EnumProperty(ObjectMappingElement, _this, "objectHandling", ObjectHandlingEnum.Create, ObjectHandlingEnum);
            /** @internal */
            _this.__objectHandlingBackup = new internal.EnumProperty(ObjectMappingElement, _this, "objectHandlingBackup", ObjectHandlingBackupEnum.Create, ObjectHandlingBackupEnum);
            /** @internal */
            _this.__isDefaultType = new internal.PrimitiveProperty(ObjectMappingElement, _this, "isDefaultType", false, internal.PrimitiveTypeEnum.Boolean);
            if (arguments.length < 4) {
                throw new Error("new ObjectMappingElement() cannot be invoked directly, please use 'model.mappings.createObjectMappingElement()'");
            }
            return _this;
        }
        Object.defineProperty(ObjectMappingElement.prototype, "containerAsMappingDocument", {
            get: function () { return _super.prototype.getContainerAs.call(this, MappingDocument); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "mappingMicroflowCall", {
            get: function () { return this.__mappingMicroflowCall.get(); },
            set: function (newValue) { this.__mappingMicroflowCall.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "children", {
            get: function () { return this.__children.get(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "entity", {
            get: function () { return this.__entity.get(); },
            set: function (newValue) { this.__entity.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "entityQualifiedName", {
            get: function () { return this.__entity.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "association", {
            get: function () { return this.__association.get(); },
            set: function (newValue) { this.__association.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "associationQualifiedName", {
            get: function () { return this.__association.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "objectHandling", {
            get: function () { return this.__objectHandling.get(); },
            set: function (newValue) { this.__objectHandling.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "objectHandlingBackup", {
            get: function () { return this.__objectHandlingBackup.get(); },
            set: function (newValue) { this.__objectHandlingBackup.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ObjectMappingElement.prototype, "isDefaultType", {
            /**
             * In version 6.4.0: introduced
             */
            get: function () { return this.__isDefaultType.get(); },
            set: function (newValue) { this.__isDefaultType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /** @internal */
        ObjectMappingElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.objectHandling = ObjectHandlingEnum.Create;
            this.objectHandlingBackup = ObjectHandlingBackupEnum.Create;
        };
        ObjectMappingElement.structureTypeName = "Mappings$ObjectMappingElement";
        ObjectMappingElement.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "isDefaultType": {
                    "introduced": "6.4.0"
                }
            }
        }, internal.StructureType.Element);
        return ObjectMappingElement;
    }(MappingElement));
    mappings.ObjectMappingElement = ObjectMappingElement;
    var ValueMappingElement = /** @class */ (function (_super) {
        __extends(ValueMappingElement, _super);
        function ValueMappingElement(model, structureTypeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, structureTypeName, id, isPartial, unit, container) || this;
            /** @internal */
            _this.__xmlDataType = new internal.PrimitiveProperty(ValueMappingElement, _this, "xmlDataType", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__type = new internal.PartProperty(ValueMappingElement, _this, "type", null, true);
            /** @internal */
            _this.__isKey = new internal.PrimitiveProperty(ValueMappingElement, _this, "isKey", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__isXmlAttribute = new internal.PrimitiveProperty(ValueMappingElement, _this, "isXmlAttribute", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__xmlPrimitiveType = new internal.EnumProperty(ValueMappingElement, _this, "xmlPrimitiveType", xmlschemas_1.xmlschemas.XmlPrimitiveType.Unknown, xmlschemas_1.xmlschemas.XmlPrimitiveType);
            /** @internal */
            _this.__isContent = new internal.PrimitiveProperty(ValueMappingElement, _this, "isContent", false, internal.PrimitiveTypeEnum.Boolean);
            /** @internal */
            _this.__attribute = new internal.ByNameReferenceProperty(ValueMappingElement, _this, "attribute", null, "DomainModels$Attribute");
            /** @internal */
            _this.__converter = new internal.ByNameReferenceProperty(ValueMappingElement, _this, "converter", null, "Microflows$Microflow");
            /** @internal */
            _this.__expectedContentTypes = new internal.PrimitiveProperty(ValueMappingElement, _this, "expectedContentTypes", "", internal.PrimitiveTypeEnum.String);
            /** @internal */
            _this.__maxLength = new internal.PrimitiveProperty(ValueMappingElement, _this, "maxLength", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__fractionDigits = new internal.PrimitiveProperty(ValueMappingElement, _this, "fractionDigits", 0, internal.PrimitiveTypeEnum.Integer);
            /** @internal */
            _this.__totalDigits = new internal.PrimitiveProperty(ValueMappingElement, _this, "totalDigits", 0, internal.PrimitiveTypeEnum.Integer);
            if (arguments.length < 4) {
                throw new Error("new ValueMappingElement() cannot be invoked directly, please use 'model.mappings.createValueMappingElement()'");
            }
            return _this;
        }
        Object.defineProperty(ValueMappingElement.prototype, "containerAsObjectMappingElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, ObjectMappingElement); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "xmlDataType", {
            /**
             * The value of this property is conceptually of type dataTypes.LegacyDataType.
             *
             * In version 7.9.0: deleted
             */
            get: function () { return this.__xmlDataType.get(); },
            set: function (newValue) { this.__xmlDataType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "type", {
            /**
             * In version 7.9.0: introduced
             */
            get: function () { return this.__type.get(); },
            set: function (newValue) { this.__type.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "isKey", {
            get: function () { return this.__isKey.get(); },
            set: function (newValue) { this.__isKey.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "isXmlAttribute", {
            get: function () { return this.__isXmlAttribute.get(); },
            set: function (newValue) { this.__isXmlAttribute.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "xmlPrimitiveType", {
            /**
             * In version 6.1.0: introduced
             */
            get: function () { return this.__xmlPrimitiveType.get(); },
            set: function (newValue) { this.__xmlPrimitiveType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "isContent", {
            get: function () { return this.__isContent.get(); },
            set: function (newValue) { this.__isContent.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "attribute", {
            get: function () { return this.__attribute.get(); },
            set: function (newValue) { this.__attribute.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "attributeQualifiedName", {
            get: function () { return this.__attribute.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "converter", {
            get: function () { return this.__converter.get(); },
            set: function (newValue) { this.__converter.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "converterQualifiedName", {
            get: function () { return this.__converter.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "expectedContentTypes", {
            /**
             * In version 6.4.1: deleted
             */
            get: function () { return this.__expectedContentTypes.get(); },
            set: function (newValue) { this.__expectedContentTypes.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "maxLength", {
            get: function () { return this.__maxLength.get(); },
            set: function (newValue) { this.__maxLength.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "fractionDigits", {
            get: function () { return this.__fractionDigits.get(); },
            set: function (newValue) { this.__fractionDigits.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ValueMappingElement.prototype, "totalDigits", {
            get: function () { return this.__totalDigits.get(); },
            set: function (newValue) { this.__totalDigits.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /** @internal */
        ValueMappingElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.fractionDigits = -1;
            this.maxLength = -1;
            this.totalDigits = -1;
            if (this.__type.isAvailable) {
                this.type = datatypes_1.datatypes.UnknownType.create(this.model);
            }
            if (this.__xmlDataType.isAvailable) {
                this.xmlDataType = "Unknown";
            }
            if (this.__xmlPrimitiveType.isAvailable) {
                this.xmlPrimitiveType = xmlschemas_1.xmlschemas.XmlPrimitiveType.Unknown;
            }
        };
        ValueMappingElement.structureTypeName = "Mappings$ValueMappingElement";
        ValueMappingElement.versionInfo = new exports.StructureVersionInfo({
            "properties": {
                "xmlDataType": {
                    "deleted": "7.9.0",
                    "deletionMessage": "Use property 'type' instead"
                },
                "type": {
                    "introduced": "7.9.0",
                    "required": {
                        "currentValue": true
                    }
                },
                "xmlPrimitiveType": {
                    "introduced": "6.1.0"
                },
                "expectedContentTypes": {
                    "deleted": "6.4.1",
                    "deletionMessage": null
                }
            }
        }, internal.StructureType.Element);
        return ValueMappingElement;
    }(MappingElement));
    mappings.ValueMappingElement = ValueMappingElement;
})(mappings = exports.mappings || (exports.mappings = {}));
var datatypes_1 = require("./datatypes");
var jsonstructures_1 = require("./jsonstructures");
var messagedefinitions_1 = require("./messagedefinitions");
var xmlschemas_1 = require("./xmlschemas");
