// mendixmodelsdk/dist/sdk/internal/transaction-checks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var deltaUtils = require("./deltas/delta-utils");
var elements_1 = require("./elements");
var TransactionChecks = /** @class */ (function () {
    function TransactionChecks(model) {
        this.model = model;
        this.checkDetachedStructures = true;
        this.detachedRequiredProperties = [];
    }
    TransactionChecks.prototype.check = function (queue) {
        if (this.checkDetachedStructures && this.model.detachedStructures.length > 0) {
            throw new Error("Some structures that were detached were not attached within the same operation");
        }
        for (var _i = 0, _a = this.detachedRequiredProperties; _i < _a.length; _i++) {
            var property = _a[_i];
            if (!property.get()) {
                throw new Error("Required part property '" + property.name + "' has null value");
            }
        }
        this.detachedRequiredProperties = [];
    };
    TransactionChecks.prototype.onNewDelta = function (delta) {
        if (delta.deltaType === "DETACH_ELEMENT") {
            var unit = deltaUtils.asModelUnit(deltaUtils.getUnitForDelta(this.model, delta));
            var element = deltaUtils.findElement(this.model, unit, delta.elementId);
            var handle = element.container._childHandle(element);
            if (handle.containingProperty.isRequired && element.container instanceof elements_1.AbstractElement) {
                this.detachedRequiredProperties.push(handle.containingProperty);
            }
        }
    };
    return TransactionChecks;
}());
exports.TransactionChecks = TransactionChecks;
